src/pg/blockchain-api/pg-blockchain-api-store.ts

```diff
@@ -0,0 +1,64 @@
+import * as postgres from 'postgres';
+import { ENV } from '../../env';
+
+export interface BlockchainDbSmartContract {
+  contract_id: string;
+  tx_id: string;
+  block_height: number;
+  abi: any;
+}
+
+export interface BlockchainDbContractLog {
+  contract_identifier: string;
+  sender_address: string;
+  value: string;
+}
+
+/**
+ * Connects and queries the Stacks Blockchain API postgres DB.
+ */
+export class PgBlockchainApiStore {
+  readonly sql: postgres.Sql<any>;
+
+  constructor() {
+    this.sql = postgres({
+      host: ENV.BLOCKCHAIN_API_PGHOST,
+      port: ENV.BLOCKCHAIN_API_PGPORT,
+      user: ENV.BLOCKCHAIN_API_PGUSER,
+      password: ENV.BLOCKCHAIN_API_PGPASSWORD,
+      database: ENV.BLOCKCHAIN_API_PGDATABASE
+    });    
+  }
+
+  async getSmartContractsCursor(
+    args: { afterBlockHeight: number }
+  ): Promise<AsyncIterable<BlockchainDbSmartContract[]>> {
+    return this.sql<BlockchainDbSmartContract[]>`
+      SELECT DISTINCT ON (contract_id) contract_id, tx_id, block_height, microblock_sequence, abi
+      FROM smart_contracts
+      WHERE
+        canonical = TRUE
+        AND microblock_canonical = TRUE
+        AND block_height >= ${args.afterBlockHeight}
+        AND abi <> '"null"'
+      ORDER BY contract_id, block_height DESC, microblock_sequence DESC
+    `.cursor();
+  }
+
+  async getContractLogsCursor(
+    args: { afterBlockHeight: number }
+  ): Promise<AsyncIterable<BlockchainDbContractLog[]>> {
+    return this.sql<BlockchainDbContractLog[]>`
+      SELECT l.contract_identifier, l.value, t.sender_address
+      FROM txs AS t 
+      INNER JOIN contract_logs AS l ON l.tx_id = t.tx_id
+      WHERE
+        t.type_id = 2
+        AND t.canonical = TRUE
+        AND t.microblock_canonical = TRUE
+        AND t.block_height >= ${args.afterBlockHeight}
+        AND l.topic = 'print'
+      ORDER BY t.block_height DESC
+    `.cursor();
```

Comment:
This is a cool use of pg cursors that looks like it would work well if we could depend on the pg instance/connection never really having downtime -- assuming there isn't any kind of "re-connect and get cursor back to original position" logic handled anywhere. 

For example, in the API, generally all of our pg queries have retry/reconnect handling code to account for this. 

If a token metadata service is started from scratch and needs to chug through the entire chain history, I could see a high possibility of something like pg restarting or dropping a connection. 

Any thoughts?

Reply:
Great point, I hadn't considered that... Just read through the postgres lib docs and it does looks like I'll need to add logic to handle that case:

> Any query which was already sent over the wire will be rejected if the connection is lost. It'll automatically defer to the error handling you have for that query, and since connections are lazy it'll automatically try to reconnect the next time a query is made

Thanks for the pointer